pkg/termination/termination.go

```diff
@@ -95,61 +95,76 @@ func (h *handler) run(ctx context.Context) error {
 	logger := h.log.WithValues("node", h.nodeName)
 	logger.V(1).Info("Monitoring node termination")
 
-	if err := wait.PollImmediateUntil(h.pollInterval, func() (bool, error) {
-		req, err := http.NewRequest("GET", h.pollURL.String(), nil)
-		if err != nil {
-			return false, fmt.Errorf("could not create request %q: %w", h.pollURL.String(), err)
+	if err := wait.PollUntilContextCancel(ctx, h.pollInterval, true, func(_ context.Context) (bool, error) {
+		terminated, err := h.checkTerminationEndpoint()
+		if !terminated {
+			logger.V(2).Info("Instance not marked for termination")
 		}
-
-		req.Header.Add("Metadata-Flavor", "Google")
-
-		resp, err := http.DefaultClient.Do(req)
-		if resp != nil {
-			defer resp.Body.Close()
-		}
-		if err != nil {
-			return false, fmt.Errorf("could not get URL %q: %w", h.pollURL.String(), err)
-		}
-
-		bodyBytes, err := ioutil.ReadAll(resp.Body)
-		if err != nil {
-			return false, fmt.Errorf("failed to read responce body: %w", err)
-		}
-
-		respBody := string(bodyBytes)
-
-		if respBody == "TRUE" {
-			// Instance marked for termination
-			return true, nil
-		}
-
-		// Instance not terminated yet
-		logger.V(2).Info("Instance not marked for termination")
-		return false, nil
-	}, ctx.Done()); err != nil {
+		return terminated, err
+	}); err != nil {
 		return fmt.Errorf("error polling termination endpoint: %w", err)
 	}
 
+	// We might arrive here due to the context being cancelled before we have gotten
+	// a clean signal from the termination endpoint, we check once more.
+	if terminated, err := h.checkTerminationEndpoint(); err != nil {
+		return err
+	} else if !terminated {
+		return nil
+	}
+
 	// Will only get here if the termination endpoint returned FALSE
 	logger.V(1).Info("Instance marked for termination, marking Node for deletion")
 
 	// Try every second to mark the node for termination up to a 30 second timeout.
 	// This should help to prevent intermittent errors and ensure we don't end up in crash loop backoff.
 	markCtx, cancel := context.WithTimeout(ctx, 30*time.Second)
 	defer cancel()
-	if err := wait.PollImmediateUntil(time.Second, func() (bool, error) {
+	if err := wait.PollUntilContextCancel(markCtx, time.Second, true, func(_ context.Context) (bool, error) {
 		if err := h.markNodeForDeletion(ctx); err != nil {
```

Comment:
This needs to use a different context I think, I expect the parent `ctx` may already be cancelled at this point

Reply:
agreed, i'll have to think this one through a little more. i suppose we could create a new background context when we go into the code path where the context is already cancelled.

Reply:
`markCtx` will also be dead by this point so that one on l121 will need some attention too